### PR TITLE
Simple GET request

### PR DESCRIPTION
**Pull request by kelseyroy**
Adds:
- Router and new router method that accepts routes.
- status_message to return the status_message associated with status_code
- HTTPServerFixture with Routes and MockRoutes that redirect to Handlers based on the associated path.
- Handler with callback function handle that returns status code and body
- SimplePost, SimpleGet, SimpleGetWithBody, TestGetWithBody handle that return a status and body based on the request method.
- New unit tests for routing functionality
- determine_routes functionality that reads the MIX_ENV and returns a test routes or default routes function

Changes: 
- parse_request and parse_headers now live in Request
- build_response, format_response, format_response headers now live in Response
- build_response is now send_resp
- how Request and Response struct types are defined

### CODE REVIEW

File: lib/request.ex [hunk 1]
```diff
@@ -1,4 +1,33 @@
...
-  @type method :: :get | :post | nil
+  defstruct method: "", path: "", resource: "", headers: %{}, body: ""
+
+  @type t :: %__MODULE__{
+          method: String.t(),
+          path: String.t(),
+          resource: String.t(),
+          headers: %{optional(String.t()) => any},
+          body: String.t()
+        }
```
Comment (Aoxian) [inline lib/request.ex]: This is pretty cool right here. I like this formatting a little better, not sure what the "best practice" is.
```suggestion
  @type t :: %__MODULE__{
    method: String.t(),
    path: String.t(),
    resource: String.t(),
    headers: %{optional(String.t()) => any},
    body: String.t()
  }
```

File: lib/response.ex [hunk 2]
```diff
@@ -1,5 +1,42 @@
...
-  @type status_message :: :ok | :not_found
+  defstruct status_code: nil, status_message: "", resource: nil, headers: %{}, body: ""
+
+  @type t :: %__MODULE__{
+          status_code: 200 | 404,
+          status_message: String.t(),
+          resource: String.t(),
+          headers: %{optional(String.t()) => any},
+          body: String.t()
+        }
```
Comment (Aoxian) [inline lib/response.ex]: I made a similar formatting comment above. 
```suggestion
  @type t :: %__MODULE__{
    status_code: 200 | 404,
    status_message: String.t(),
    resource: String.t(),
    headers: %{optional(String.t()) => any},
    body: String.t()
  }
```

File: lib/response.ex [hunk 3]
```diff
@@ -1,5 +1,42 @@
...
+    "#{res.resource} #{res.status_code} #{res.status_message}" <>
+      carriage_return <>
+      "#{format_response_headers(res.headers)}" <>
+      carriage_return <>
+      "#{res.body}"
+  end
+
+  defp format_response_headers(headers) do
+    for {key, val} <- headers, into: "", do: "#{key}: #{val}\r\n"
+  end
```
Comment (Aoxian) [inline lib/response.ex]: Make the `carriage_return` variable a constant in the class? Then use it in both methods on this class?
```suggestion
  defp format_response_headers(headers) do
    for {key, val} <- headers, into: "", do: "#{key}: #{val}#{carriage_return}"
  end
```

Comment (Aoxian) [reply lib/response.ex]: This particular constant is used all over the place. Is there an easy way to declare it and use it?

File: lib/router.ex [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  alias HTTPServer.Response
+  alias HTTPServer.Request
+
+  def route(%Request{method: "POST", path: "/echo_body"} = req) do
+    Response.build_response(req, 200)
+  end
+
+  def route(%Request{method: "GET", path: "/simple_get"} = req) do
+    Response.build_response(req, 200)
+  end
```
Comment (Aoxian) [inline lib/router.ex]: I think there is a potential bug here. If your `req` has a body, it will be echoed because of how `build_response` is defined on the `Response`

File: lib/response.ex [hunk 5]
```diff
@@ -1,5 +1,42 @@
...
+
+  def build_response(req, status_code) do
+    %__MODULE__{
+      status_code: status_code,
+      status_message: status_message(status_code),
+      resource: req.resource,
+      headers: req.headers,
+      body: req.body
+    }
+  end
```
Comment (Aoxian) [inline lib/response.ex]: I would expect a response builder to build a response based off of a Response object.  This takes a `req`, that is a bit code smelly.

Something more like this, which also guarantees that `build_response` only works on `%Response`-s
```suggestion
  def build_response(response = %Response{}, status_code) do
    %__MODULE__{
      status_code: status_code,
      status_message: status_message(status_code),
      resource: response.resource,
      headers: response.headers,
      body: response.body
    }
  end
```

File: lib/request.ex (hunk 1 above)
Comment (kelseyroy) [reply lib/request.ex]: I like this better too. A lot of the examples I've seen put the [struct literal syntax all on one line](https://hexdocs.pm/elixir/Kernel.html#defstruct/1-types), but I like the returns between and your proposed spacing looks neater. I've been thinking about trying to add [documentation to my modules](https://elixirschool.com/en/lessons/basics/documentation#best-practice-7), but that feels low in priority to getting it working.

File: lib/response.ex (hunk 5 above)
Comment (Aoxian) [reply lib/response.ex]: In other words, should a `%Response` know about a `%Request` at all?

File: lib/router.ex [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+  def route(%Request{method: "GET", path: "/simple_get_with_body"} = req) do
+    body = "Hello world"
+    req_with_body = %{req | body: body, headers: %{req.headers | "Content-Length" => "#{String.length(body)}"}}
+    Response.build_response(req_with_body, 200)
+  end
```
Comment (Aoxian) [inline lib/router.ex]: This "works" but is it "right"?  This router doesn't seem to route at all, it appears to build responses.

File: test/http_server_spec/server.rb [hunk 7]
```diff
@@ -30,7 +30,7 @@
 end
 
 get "/simple_get_with_body" do
-  "Hello world"
+  request.body
```
Comment (Aoxian) [inline test/http_server_spec/server.rb]: Is this change necessary?  I think for as much as you can, do not touch the `http_server_spec` code.  Only remove the `@wip`-s as you get stories to pass.

File: lib/router.ex (hunk 6 above)
Comment (Aoxian) [reply lib/router.ex]: And specifically responses that satisfy the `http_server_spec` integration test suite.  How can you decouple that, so that you can build out generic reusable functionality, while still having a suite of integration tests that test that reusable functionality?

File: test/http_server_test.exs [hunk 8]
```diff
@@ -1,15 +1,38 @@
...
+        resource: "HTTP/1.1",
+        headers: %{
+          "Accept" => "*/*",
+          "Content-Length" => "9",
+          "Content-Type" => "text/plain",
+          "Host" => "127.0.0.1 4000",
+          "User-Agent" => "ExampleBrowser/1.0"
+        },
+        body: ""
+      }
```
Comment (Aoxian) [inline test/http_server_test.exs]: There is already an integration test to test this specific `/simple_get` endpoint.  To keep from coupling your http_server implementation to the `http_server_spec` tests, I suggest not re-using the endpoint names in the unit tests.  

I bet this breaks if you change `GET /simple_get HTTP/1.1\r\n` to `GET /test_get HTTP/1.1\r\n`.  Try mocking out the declaration of that test endpoint (`/test_get`), it will open up a world of missing design here.

**Review by Aoxian — CHANGES_REQUESTED** [review]
I think the unit tests, are too tightly coupled to the `http_server_spec` tests.  I suggest mocking out test endpoints that are different than what the `http_server_spec` is using. 

secondly (optionally - it could come later) the `router` should route to *something* instead of building the responses.

File: test/http_server_spec/server.rb (hunk 7 above)
Comment (kelseyroy) [reply test/http_server_spec/server.rb]: This was mistakenly left in from when I was worried that the acceptance tests were preventing the test from passing! I'll change that back right now.

File: lib/router.ex (hunk 6 above)
Comment (kelseyroy) [reply lib/router.ex]: I see what you're saying, and I think I was questioning the implementation myself. I'm changing this now so that the only thing passed to build_response is the status code, and potentially any body that should be included as data in the response (almost like how you use res.send() to send your response in express.js).

File: lib/response.ex (hunk 5 above)
Comment (kelseyroy) [reply lib/response.ex]: Now that you point it out, I agree that it shouldn't. I mentioned below but I'm changing this now so that the only thing passed to build_response is the status code and a maybe a body. I think I'm realizing too that the headers and resource don't necessarily have to be exact replica's of what was sent in the request, they can convey any messages they want as long as it's clear what they mean to the person making the request.

File: test/http_server_test.exs (hunk 8 above)
Comment (kelseyroy) [reply test/http_server_test.exs]: Ok great, I'll make these updates to the tests!

File: lib/router.ex (hunk 6 above)
Comment (Aoxian) [reply lib/router.ex]: If you are going to introduce the concept of a router, is now also the time to introduce the concepts of handler and/or controllers? Does it make sense to route without them?

File: lib/request.ex [hunk 9]
```diff
@@ -1,4 +1,33 @@
...
+  @type t :: %__MODULE__{
+    method: String.t(),
+    path: String.t(),
+    resource: String.t(),
+    headers: %{optional(String.t()) => any},
+    body: String.t()
+  }
+
+  def parse_request(message) do
+    [request_data | body] = message |> String.split("\r\n\r\n")
```
Comment (t-keazirian) [inline lib/request.ex]: [Carriage return](https://github.com/kelseyroy/http_server/pull/6/files#r1220290460) here too 👀

File: lib/http_server.ex [hunk 10]
```diff
@@ -50,58 +51,4 @@ defmodule HTTPServer do
     :gen_tcp.send(socket, data)
     socket
   end
-
```
Comment (t-keazirian) [inline lib/http_server.ex:53]: nice clean-up here! ✂️

File: lib/router.ex (hunk 6 above)
Comment (kelseyroy) [reply lib/router.ex]: I was looking into that a little bit yesterday evening, I think it felt unclear because there isn't a "source" of data that I would query in this instance. But I'm realizing that perhaps it could exist in the current state to just return a string and send it to the response? Or should I maybe create a file that's just a JSON object with some of the necessary "data" aka return strings that I can read from?

**Comment by kelseyroy** [comment]
[Simple GET request](https://trello.com/c/QGDkwxmi/15-simple-get-request)

File: lib/response.ex (hunk 3 above)
Comment (kelseyroy) [reply lib/response.ex]: For now I've added it as a module attribute to the modules that use it (Request, Response, and the HTTPServerTest) but I would like to make it even more global than this going forward somehow.

File: lib/request.ex (hunk 9 above)
Comment (kelseyroy) [reply lib/request.ex]: > For now I've added it as a module attribute to the modules that use it (Request, Response, and the HTTPServerTest)...

I mentioned this to Ian up above as well, I'm going to look into how I can make it even more global than this!

File: lib/router.ex (hunk 6 above)
Comment (kelseyroy) [reply lib/router.ex]: This has been updated btw! Let me know if I'm close to getting it right. I think there might still be some flaws in the logic of how I've set it up, but I do think I've created greater separation between the router's responsibility and what the acceptance tests need to pass.

File: test/http_server_test.exs (hunk 8 above)
Comment (kelseyroy) [reply test/http_server_test.exs]: The unit tests have been updated and now live in HTTPServerTest.Router. Let me know if you think I could go a bit further to decouple them from the integration test functionality.

File: lib/http_server_fixture/mock_routes.ex [hunk 11]
```diff
@@ -0,0 +1,8 @@
+defmodule HTTPServer.MockRoutes do
+  alias HTTPServer.Request
+  alias HTTPServerFixture.Handler
+
+  def route(%Request{path: "/post_test"} = req), do: Handler.SimplePost.handle(req)
+  def route(%Request{path: "/get_test"} = req), do: Handler.SimpleGet.handle(req)
+  def route(%Request{path: "/get_test_with_body"} = req), do: Handler.TestGetWithBody.handle(req)
+end
```
Comment (Aoxian) [inline lib/http_server_fixture/mock_routes.ex]: these look like unit test mocks, they should probably be separated from the `http_server_fixture`(because that's for integration tests).  I suggest maybe getting a little more creative with folder structure to help organize that.  Typically all the unit test related mocks / supporting files all go in the `/test` folder for Elixir projects.

Comment (Aoxian) [reply lib/http_server_fixture/mock_routes.ex]: for unit tests, you shouldn't have to actually start the server, where for the fixture you would. So the fixture code living in `/lib` makes sense to me.

File: lib/http_server_fixture/handler/handler.ex [hunk 12]
```diff
@@ -0,0 +1,4 @@
+defmodule HTTPServerFixture.Handler do
+  alias HTTPServer.Request
+  @callback handle(req :: %Request{}) :: {status_code :: integer(), body :: term()}
+end
```
Comment (Aoxian) [inline lib/http_server_fixture/handler/handler.ex]: this is actual http_server code. I suggest placing it in the `HTTPServer` module, and not the fixture.

Looks great though!

File: lib/http_server.ex [hunk 13]
```diff
@@ -21,12 +23,12 @@ defmodule HTTPServer do
...
       socket
       |> read_line()
-      |> parse_request()
+      |> Request.parse_request()
 
     response =
       request
-      |> build_response()
-      |> format_response()
+      |> Router.router(&(Routes.route(&1)))
```
Comment (Aoxian) [inline lib/http_server.ex]: this is Elixry to the point of confusion.  I see that production code will run the fixture Routes.  Is it possible to only use those routes when starting a server to integration tests? and use a default Routes for when you start the server with production code?

File: lib/http_server_fixture/mock_routes.ex (hunk 11 above)
Comment (kelseyroy) [reply lib/http_server_fixture/mock_routes.ex]: They were originally located in the test directory as a HTTPServerTest module, but it kept throwing an error that it couldn't access HTTPServerFixture.Handler when they were there. I'll keep playing around with it to see if there was something else that was preventing them from reaching it!

File: lib/http_server.ex (hunk 13 above)
Comment (kelseyroy) [reply lib/http_server.ex]: Ooooh great idea. The only thing that could be challenging is that the server has to be started before the integration tests are called, so I may need to make a specific start task that spins up the server + routes for the integration test suite?

File: lib/http_server_fixture/mock_routes.ex (hunk 11 above)
Comment (Aoxian) [reply lib/http_server_fixture/mock_routes.ex]: [this comment.](https://github.com/kelseyroy/http_server/pull/6#discussion_r1223381592) is related.

File: lib/http_server.ex (hunk 13 above)
Comment (Aoxian) [reply lib/http_server.ex]: I like that idea, can't wait to see if it works

Comment (kelseyroy) [reply lib/http_server.ex]: Let me know if the new `determine_routes` functionality is what you imagined here. I feel like there is probably a better way, but after a lot of trial and error I found this to be the easiest and quickest fix!

File: lib/http_server_fixture/handler/handler.ex (hunk 12 above)
Comment (kelseyroy) [reply lib/http_server_fixture/handler/handler.ex]: `handle` is now in the `HTTPServer.Handler` module!

File: test/http_router_test.exs [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  alias HTTPServer.Router
+  alias HTTPServer.Request
+  alias HTTPServer.Response
+  # doctest HTTPServer
+
+  defmodule HttpRouterTest.MockRoutes do
+    alias HTTPServer.Request
+    alias HTTPServerFixture.SimpleGet
+    alias HTTPServerFixture.TestGetWithBody
+    alias HTTPServerFixture.SimplePost
```
Comment (Aoxian) [inline test/http_router_test.exs]: These are still `HTTPServerFixture` code.  You could just make mock routes right in this file to start.

```elixir
defmodule HTTPServer.TestGetHandler do
  alias HTTPServer.Request
  @behaviour HTTPServer.Handler

  @impl HTTPServer.Handler
  def handle(%Request{method: "GET"} = _req) do
    {200, "this is a mocked get request"}
  end
end
```

File: lib/http_server.ex [hunk 15]
```diff
@@ -51,57 +53,12 @@ defmodule HTTPServer do
...
-
-    case method do
-      "POST" ->
-        %{req | method: :post}
-
-      _ ->
-        {:error, :unknown_command}
+    case env do
+      :test -> &HTTPServerFixture.Routes.route/1
+      _ -> &Routes.route/1
```
Comment (Aoxian) [inline lib/http_server.ex]: I like the creativity of this solution.  though I would point out that there should be an explicit difference between something like `:test` (which connotes unit tests) and `:integration_test` (which would be the http_server_specs).
```suggestion
    case env do
      :test -> &HTTPServer.TestRoutes.route/1
      :integration_test -> &HTTPServerFixture.Routes.route/1
      _ -> &Routes.route/1
```

Comment (Aoxian) [reply lib/http_server.ex]: also, there appears to be configuration code here, I wonder if this is the best place for this logic?  I am not really sure, so that's food for thought for a future PR.

File: lib/http_server.ex [hunk 16]
```diff
@@ -1,6 +1,8 @@
 defmodule HTTPServer do
   alias HTTPServer.Response
   alias HTTPServer.Request
+  alias HTTPServer.Router
+  alias HTTPServer.Routes
```
Comment (Aoxian) [inline lib/http_server.ex]: nice to see there are no HTTPServerFixture dependencies here.

File: test/http_router_test.exs [hunk 17]
```diff
@@ -0,0 +1,113 @@
+defmodule HTTPServerTest.Router do
+  use ExUnit.Case
+  alias HTTPServer.Router
+  alias HTTPServer.Request
+  alias HTTPServer.Response
+  # doctest HTTPServer
```
Comment (Aoxian) [inline test/http_router_test.exs]: 🔫 🔫 🔫

File: lib/http_server_fixture/handler/handler.ex (hunk 12 above)
Comment (Aoxian) [reply lib/http_server_fixture/handler/handler.ex]: 🎉

File: lib/http_server.ex (hunk 13 above)
Comment (Aoxian) [reply lib/http_server.ex]: Its super creative, I left a comment, but I like it!

File: lib/http_server_handler/routes.ex [hunk 18]
```diff
@@ -0,0 +1,7 @@
+defmodule HTTPServer.Routes do
+  alias HTTPServer.Request
+  # alias HTTPServer.Handler
```
Comment (t-keazirian) [inline lib/http_server_handler/routes.ex]: nit: remove commented code if no longer needed 😄

**Review by Aoxian — APPROVED** [review]

File: test/http_server_test.exs [hunk 19]
```diff
@@ -33,50 +28,26 @@ defmodule HTTPServerTest do
           "Host" => "127.0.0.1 4000",
           "User-Agent" => "ExampleBrowser/1.0"
         },
-        body: "some body"
+        body: ""
       }
 
-      assert HTTPServer.parse_request(message) == expected_parsed_request
+      assert Request.parse_request(message) == expected_parsed_request
```
Comment (Aoxian) [inline test/http_server_test.exs:34]: something I noticed about these "`http_server_test`" tests, they are actually testing the `HTTPServer.Request.parse_request/1` function.

Since `HTTPServer` is "main", you don't really need to test "main"

I think this file should be moved to reflect what its actually unit testing, `HTTPServer.Request` aka `request_test.exs`

Comment (Aoxian) [reply test/http_server_test.exs:34]: I wonder what other unit tests could be added as well? (example: `response_test.exs`)

Comment (kelseyroy) [reply test/http_server_test.exs:34]: I like this idea! But I might fold them into the Simple HEAD ticket since this PR is pretty big.

File: lib/http_server_handler/routes.ex (hunk 18 above)
Comment (kelseyroy) [reply lib/http_server_handler/routes.ex]: Great catch! This was also being flagged as a problem (even though it wasn't being used?) so I'm glad you pointed it out!
